File: scripts/enrich-members-snapshot.py
#!/usr/bin/env python3
"""
Enrich members snapshot with survey scores from Excel file
Usage: python3 scripts/enrich-members-snapshot.py
"""

import json
import zipfile
import xml.etree.ElementTree as ET
import sys
from pathlib import Path

def normalize_text(value):
    """Normalize text for matching"""
    if not value:
        return ''
    return str(value).strip().upper()

def extract_surname(full_name):
    """Extract surname from full name"""
    normalized = normalize_text(full_name)
    parts = normalized.split()
    return parts[-1] if parts else ''

def extract_excel_data(xlsx_path, sheet_num=2):
    """Extract survey data from Excel file"""
    data = []
    
    with zipfile.ZipFile(xlsx_path, 'r') as zip_ref:
        sheet_path = f'xl/worksheets/sheet{sheet_num}.xml'
        try:
            sheet_xml = zip_ref.read(sheet_path).decode('utf-8')
        except:
            print(f"Error: Sheet {sheet_num} not found", file=sys.stderr)
            return data
        
        root = ET.fromstring(sheet_xml)
        
        # Extract shared strings
        strings = []
        try:
            strings_xml = zip_ref.read('xl/sharedStrings.xml').decode('utf-8')
            strings_root = ET.fromstring(strings_xml)
            for si in strings_root.findall('{http://schemas.openxmlformats.org/spreadsheetml/2006/main}si'):
                text_parts = []
                for t in si.findall('.//{http://schemas.openxmlformats.org/spreadsheetml/2006/main}t'):
                    if t.text:
                        text_parts.append(t.text)
                strings.append(''.join(text_parts))
        except:
            pass
        
        # Parse cells
        rows = []
        for row in root.findall('{http://schemas.openxmlformats.org/spreadsheetml/2006/main}sheetData/{http://schemas.openxmlformats.org/spreadsheetml/2006/main}row'):
            row_data = []
            for cell in row.findall('{http://schemas.openxmlformats.org/spreadsheetml/2006/main}c'):
                cell_type = cell.get('t', 'n')
                value_elem = cell.find('{http://schemas.openxmlformats.org/spreadsheetml/2006/main}v')
                
                if value_elem is not None and value_elem.text:
                    if cell_type == 's':
                        idx = int(value_elem.text)
                        value = strings[idx] if idx < len(strings) else ''
                    else:
                        try:
                            value = float(value_elem.text)
                        except:
                            value = value_elem.text
                    row_data.append(value)
                else:
                    row_data.append('')
            
            if any(row_data):
                rows.append(row_data)
        
        # Convert to dict format
        if len(rows) > 1:
            headers = rows[0]
            for row in rows[1:]:
                entry = {}
                for i, header in enumerate(headers):
                    if isinstance(header, str):
                        header = header.strip()
                    entry[header] = row[i] if i < len(row) else ''
                if any(entry.values()):
                    data.append(entry)
    
    return data

def build_survey_map(survey_data):
    """Build lookup maps for survey data"""
    by_display_name = {}
    by_surname_seat = {}
    
    for entry in survey_data:
        name = entry.get('Name', '')
        if not name:
            continue
        
        # By exact name
        name_key = normalize_text(name)
        if name_key:
            by_display_name[name_key] = entry
        
        # By surname + constituency
        surname = extract_surname(name)
        
        # Current parliament constituency
        const_name = entry.get('Constituency (2024- parliament)', '')
        if surname and const_name:
            const_key = normalize_text(const_name)
            by_surname_seat[f'{surname}|{const_key}'] = entry
        
        # Old parliament constituency (fallback)
        const_name_old = entry.get('Constituency (2019-2024 parliament)', '')
        if surname and const_name_old:
            const_key_old = normalize_text(const_name_old)
            by_surname_seat[f'{surname}|{const_key_old}'] = entry
    
    return by_display_name, by_surname_seat

def find_survey_entry(member, by_display_name, by_surname_seat):
    """Find survey entry for a member"""
    # Try exact name match
    exact_match = by_display_name.get(normalize_text(member['name']))
    if exact_match:
        return exact_match
    
    # Try surname + membership constituency
    surname = extract_surname(member['name'])
    seat_name = member.get('membershipFrom', '')
    if surname and seat_name:
        seat_key = normalize_text(seat_name)
        match = by_surname_seat.get(f'{surname}|{seat_key}')
        if match:
            return match
    
    return None

def enrich_member(member, by_display_name, by_surname_seat):
    """Add survey scores to a member"""
    survey_entry = find_survey_entry(member, by_display_name, by_surname_seat)
    
    if survey_entry:
        value = survey_entry.get('Value')
        value_float = float(value) if value else None
        # Get constituency code from current parliament (2024) or fallback to 2019
        const_code = survey_entry.get('Constituency code (2024- parliament)', '') or survey_entry.get('Constituency code (2019-2024 parliament)', '')
        
        return {
            **member,
            'value': value_float,  # 0-100 left-right score
            'value_lo': float(survey_entry.get('Could be as low as', 0)) or None,
            'value_hi': float(survey_entry.get('Could be as high as', 100)) or None,
            # For compatibility with old code
            'econ_mean': value_float,
            'cult_mean': value_float,
            'constituencyCode': const_code
        }
    else:
        return {
            **member,
            'value': None,
            'value_lo': None,
            'value_hi': None,
            'econ_mean': None,
            'cult_mean': None,
            'constituencyCode': None
        }

def main():
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    snapshot_path = project_root / 'data' / 'members-snapshot.json'
    
    print(f'Loading snapshot from {snapshot_path}...', file=sys.stderr)
    snapshot = json.loads(snapshot_path.read_text())
    members = snapshot.get('members', [])
    print(f'Found {len(members)} members', file=sys.stderr)
    
    # Extract Excel data
    print('Extracting survey data from Excel...', file=sys.stderr)
    survey_data = extract_excel_data('/tmp/survey.xlsx', sheet_num=2)
    print(f'Loaded {len(survey_data)} survey entries', file=sys.stderr)
    
    # Build lookup maps
    by_display_name, by_surname_seat = build_survey_map(survey_data)
    print(f'Built maps: {len(by_display_name)} by name, {len(by_surname_seat)} by surname+seat', file=sys.stderr)
    
    # Enrich members
    print('Enriching members with survey scores...', file=sys.stderr)
    enriched = [enrich_member(m, by_display_name, by_surname_seat) for m in members]
    
    # Count enriched
    with_scores = sum(1 for m in enriched if m.get('value') is not None)
    print(f'Enriched {with_scores} of {len(enriched)} members with scores', file=sys.stderr)
    
    # Verify target MPs
    print('\nTarget MPs:', file=sys.stderr)
    for name in ['Josh Babarinde', 'Jade Botterill', 'Helena Dollimore']:
        member = next((m for m in enriched if m['name'] == name), None)
        if member:
            print(f"  {name}: value={member.get('value')}", file=sys.stderr)
    
    # Save enriched snapshot
    enriched_snapshot = {
        'generatedAt': __import__('datetime').datetime.now(
__import__('datetime').timezone.utc).isoformat(),
        'members': enriched
    }
    
    snapshot_path.write_text(json.dumps(enriched_snapshot, indent=2))
    print(f'\nSaved enriched snapshot to {snapshot_path}', file=sys.stderr)

if __name__ == '__main__':
    main()

File: scripts/enrich-members-snapshot.js
#!/usr/bin/env node
/**
 * Enrich members snapshot with survey scores from the CSV
 * Usage: node scripts/enrich-members-snapshot.js
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const snapshotPath = path.join(__dirname, '../data/members-snapshot.json');

// Normalization functions (same as in app.js)
function normalizeText(value) {
  return (value || '').toString().trim().toUpperCase()
    .replace(/[^A-Z0-9]+/g, ' ').replace(/\s+/g, ' ').trim();
}

function extractSurname(fullName) {
  const normalized = normalizeText(fullName);
  const parts = normalized.split(' ').filter(Boolean);
  return parts.length ? parts[parts.length - 1] : '';
}

// Parse CSV text
function parseCsv(text) {
  const rows = [];
  let current = '';
  let row = [];
  let inQuotes = false;
  
  for (let i = 0; i < text.length; i++) {
    const char = text[i];
    if (inQuotes) {
      if (char === '"') {
        if (text[i + 1] === '"') { 
          current += '"'; 
          i += 1; 
        } else { 
          inQuotes = false; 
        }
      } else {
        current += char;
      }
    } else if (char === '"') {
      inQuotes = true;
    } else if (char === ',') {
      row.push(current); 
      current = '';
    } else if (char === '\r') {
      continue;
    } else if (char === '\n') {
      row.push(current); 
      rows.push(row); 
      current = ''; 
      row = [];
    } else {
      current += char;
    }
  }
  
  if (current !== '' || row.length) { 
    row.push(current); 
    rows.push(row); 
  }
  
  if (!rows.length) return [];
  const headers = rows.shift().map(h => h.trim());
  return rows.map(values => {
    const obj = {};
    headers.forEach((h, i) => { obj[h] = values[i] ?? ''; });
    return obj;
  });
}

async function fetchCsv(url) {
  console.log(`Fetching survey data from ${url}...`);
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch CSV: ${response.status}`);
  }
  return response.text();
}

async function buildSurveyMap(csvText) {
  const entries = parseCsv(csvText);
  const byDisplayName = new Map();
  const bySurnameSeat = new Map();
  
  entries.forEach(entry => {
    const displayNameKey = normalizeText(entry.DisplayName);
    if (displayNameKey) byDisplayName.set(displayNameKey, entry);
    
    const surnameKey = extractSurname(entry.DisplayName);
    const seatNameKey = normalizeText(entry.PCON24NM || entry.PCON22NM || '');
    const seatCodeKey = normalizeText(entry.PCON24CD || entry.PCON22CD || '');
    
    if (surnameKey && seatCodeKey) {
      bySurnameSeat.set(`${surnameKey}|${seatCodeKey}`, entry);
    }
    if (surnameKey && seatNameKey) {
      bySurnameSeat.set(`${surnameKey}|${seatNameKey}`, entry);
    }
  });
  
  return { byDisplayName, bySurnameSeat };
}

function findSurveyEntry(member, surveyMap) {
  // Try exact name match
  const exactMatch = surveyMap.byDisplayName.get(normalizeText(member.name));
  if (exactMatch) return exactMatch;
  
  // Try surname + seat name
  const surname = extractSurname(member.name);
  const seatName = normalizeText(member.membershipFrom || '');
  if (surname && seatName) {
    const nameMatch = surveyMap.bySurnameSeat.get(`${surname}|${seatName}`);
    if (nameMatch) return nameMatch;
  }
  
  return null;
}

function enrichMember(member, surveyMap) {
  const surveyEntry = findSurveyEntry(member, surveyMap);
  
  if (surveyEntry) {
    return {
      ...member,
      econ_mean: surveyEntry.econ_mean ? parseFloat(surveyEntry.econ_mean) : null,
      econ_lo: surveyEntry.econ_lo ? parseFloat(surveyEntry.econ_lo) : null,
      econ_hi: surveyEntry.econ_hi ? parseFloat(surveyEntry.econ_hi) : null,
      cult_mean: surveyEntry.cult_mean ? parseFloat(surveyEntry.cult_mean) : null,
      cult_lo: surveyEntry.cult_lo ? parseFloat(surveyEntry.cult_lo) : null,
      cult_hi: surveyEntry.cult_hi ? parseFloat(surveyEntry.cult_hi) : null
    };
  }
  
  return {
    ...member,
    econ_mean: null,
    econ_lo: null,
    econ_hi: null,
    cult_mean: null,
    cult_lo: null,
    cult_hi: null
  };
}

async function main() {
  try {
    // Load current snapshot
    console.log('Loading current snapshot...');
    const snapshot = JSON.parse(fs.readFileSync(snapshotPath, 'utf-8'));
    const members = snapshot.members || [];
    console.log(`Found ${members.length} members in snapshot`);
    
    // Fetch survey data
    let csvText;
    try {
      csvText = await fetchCsv('https://raw.githubusercontent.com/chrishanretty/pairwise_mps_2024/main/outputs/summary_statistics_file.csv');
    } catch {
      console.log('Primary URL failed, trying CDN fallback...');
      csvText = await fetchCsv('https://cdn.jsdelivr.net/gh/chrishanretty/pairwise_mps_2024@main/outputs/summary_statistics_file.csv');
    }
    
    console.log(`Fetched CSV: ${csvText.length} bytes`);
    
    // Build survey map
    console.log('Building survey data index...');
    const surveyMap = await buildSurveyMap(csvText);
    console.log(`Survey map: ${surveyMap.byDisplayName.size} by display name, ${surveyMap.bySurnameSeat.size} by surname+seat`);
    
    // Enrich members
    console.log('Enriching members with survey scores...');
    const enriched = members.map(member => enrichMember(member, surveyMap));
    
    // Count enriched members
    const withScores = enriched.filter(m => m.econ_mean !== null || m.cult_mean !== null).length;
    console.log(`Enriched ${withScores} of ${enriched.length} members with survey scores`);
    
    // Check specific MPs
    const babarinde = enriched.find(m => m.name === 'Josh Babarinde');
    const botterill = enriched.find(m => m.name === 'Jade Botterill');
    const dollimore = enriched.find(m => m.name === 'Helena Dollimore');
    
    console.log('\nTarget MPs:');
    console.log(`- Josh Babarinde: econ=${babarinde?.econ_mean}, cult=${babarinde?.cult_mean}`);
    console.log(`- Jade Botterill: econ=${botterill?.econ_mean}, cult=${botterill?.cult_mean}`);
    console.log(`- Helena Dollimore: econ=${dollimore?.econ_mean}, cult=${dollimore?.cult_mean}`);
    
    // Save enriched snapshot
    const enrichedSnapshot = {
      generatedAt: new Date().toISOString(),
      members: enriched
    };
    
    fs.writeFileSync(snapshotPath, JSON.stringify(enrichedSnapshot, null, 2));
    console.log(`\nSaved enriched snapshot to ${snapshotPath}`);
    
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();
